src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py

```diff
@@ -0,0 +1,162 @@
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author S. Siso, STFC Daresbury Lab
+
+'''Module containing tests for the FoldConditionalReturnExpressionsTrans
+transformation.'''
+
+from __future__ import absolute_import
+import pytest
+
+from fparser.common.readfortran import FortranStringReader
+from psyclone.psyir.transformations import \
+        FoldConditionalReturnExpressionsTrans
+from psyclone.psyir.backend.fortran import FortranWriter
+from psyclone.psyir.frontend.fparser2 import Fparser2Reader
+from psyclone.psyir.transformations import TransformationError
+
+
+def test_description():
+    ''' Check that the transformation returns the expected strings '''
+    trans = FoldConditionalReturnExpressionsTrans()
+    assert trans.name == "FoldConditionalReturnExpressionsTrans"
+    assert str(trans) == ("Re-structure kernel statements to eliminate "
+                          "conditional Return expressions.")
+
+
+def test_validation():
+    ''' Check that the transformation can only be applied to routine nodes '''
+    trans = FoldConditionalReturnExpressionsTrans()
+    with pytest.raises(TransformationError) as info:
+        trans.apply(None)
+    assert("Error in FoldConditionalReturnExpressionsTrans transformation. "
+           "This transformation can only be applied to 'Routine' nodes, but "
+           "found 'NoneType'." in str(info.value))
+
+
+SUB_IN1 = (
+    "subroutine sub1(i, a)\n"
+    "  real, intent(inout) :: a\n"
+    "  integer, intent(in) :: i\n"
+    "  if (i < 5) then\n"
+    "    return\n"
+    "  endif\n"
+    "  if (i > 10) then\n"
+    "    ! Comments do not matter\n"
+    "    return\n"
+    "    a=2.0 ! Dead code does not matter\n"
+    "  endif\n"
+    "  a=0.0\n"
+    "  a=1.0\n"
+    "  a=2.0\n"
+    "end subroutine\n")
+SUB_OUT1 = (
+    "subroutine sub1(i, a)\n"
+    "  real, intent(inout) :: a\n"
+    "  integer, intent(in) :: i\n\n"
+    "  if (.NOT.i < 5) then\n"
+    "    if (.NOT.i > 10) then\n"
+    "      a = 0.0\n"
+    "      a = 1.0\n"
+    "      a = 2.0\n"
+    "    end if\n"
+    "  end if\n\n"
+    "end subroutine sub1\n")
+
+# Tests with preceding code that is not part of the mask
+SUB_IN2 = (
+    "subroutine sub1(i, a)\n"
+    "  real, intent(inout) :: a\n"
+    "  integer, intent(in) :: i\n"
+    "  {0}\n"
+    "  if (i < 5) then\n"
+    "    return\n"
+    "  endif\n"
+    "  if (i > 10) then\n"
+    "    return\n"
+    "  endif\n"
+    "  a=0.0\n"
+    "end subroutine\n")
+SUB_IN2_1 = SUB_IN2.format("a = 0.0")
+SUB_IN2_2 = SUB_IN2.format(
```

Comment:
Please add comments in here detailing why each case will be ignored by the transformation.

Reply:
Updated the comment at the top. Is that enough?

Reply:
Yes, I think it's clear now, thanks.